### ISSUE TITLE
Faster cluster startup and shutdown

### ISSUE THREAD

**Issue by pashkinelfe** [orioledb/orioledb#328]
With on disk db cluster size of ~25Gb (10000 partitions), s3 bucket: 
<img width="1052" alt="image" src="https://github.com/orioledb/orioledb/assets/63344111/8d764d5c-63e6-413b-bb26-7211bbb1c5db">

Everything works fine.

pg-ctl start (and stop) take around 4.5 minutes each (without specifying -t it will report timeout error, the cluster starts after the error anyway)

Almost all this time at startup log is full of messages:
`DEBUG:  initial_parts_counting_callback`

At stop messages:
`DEBUG:  eviction_callback`

Could any optimizations make these operations a little bit faster?

**Comment by pashkinelfe** [comment]
- With 10000 partitions and ~25 Gb cluster it takes ~4.5 minutes to start/stop cluster for both s3 and non-s3 mode
- With 100 partitions and ~25 Gb it takes 10-20 seconds to start/stop cluster for both s3 and non-s3 mode

**Comment by akorotkov** [comment]
Thank you, @pashkinelfe. Could you, please, make a flamegraph?

**Comment by pashkinelfe** [comment]
Non-s3 mode. Database 6Gb (I limited it, otherwise it's slow to fill). 10000 partitions

Shutdown (5 min) takes 2 long checkpoints (checkpointer process @ ~25% CPU):
First checkpoint:
![shutdown-10000-checkpoint1](https://github.com/orioledb/orioledb/assets/63344111/e0d7e06f-2ff1-425e-b93e-e55f593e96b9)
Second checkpoint
![shutdown-10000-checkpoint2](https://github.com/orioledb/orioledb/assets/63344111/002c805e-2961-46a8-b9c4-4d9f32d21896)

Startup (1.3 min) (main postgres process @ ~2% CPU)
![startup-1000](https://github.com/orioledb/orioledb/assets/63344111/4f1b2ccc-50a3-4f78-95b3-25a690f70d9e)

**Comment by pashkinelfe** [comment]
S3 mode. Database 6Gb. 10000 partitions

Shutdown ~5min: checkpointer @50% CPU + many s3 workers @ ~20% CPU each
![shutdown-s3-10000-checkpointer](https://github.com/orioledb/orioledb/assets/63344111/c290aed7-10b2-4919-bbcb-42155e968211)
![shutdown-s3-10000-s3worker](https://github.com/orioledb/orioledb/assets/63344111/b76153ea-c1a8-4025-8a6d-436c2f660218)


Startup: 
1 min counting partitions @100% CPU (part1)
5 min cleanup old file @ 2% CPU (part2) - this is similar to non-s3 mode but takes ~3 times longer.
![startup-s3-10000-part1](https://github.com/orioledb/orioledb/assets/63344111/8123ecbf-12d3-4e2f-90c6-104a5c803a80)
![startup-s3-10000-part2](https://github.com/orioledb/orioledb/assets/63344111/22a034ca-d38d-48ea-87ca-5ddd1f9bf794)

Probes taken: 1000 per second for 60 seconds.
(non-recompressed svg's 
[s3-10000.zip](https://github.com/orioledb/orioledb/files/13779186/s3-10000.zip)
)

**Comment by akorotkov** [comment]
Please, recheck on main (234115c)

**Comment by pashkinelfe** [comment]
On the same s3 db as previous measurements:
Startup:
53s counting partitions @100% CPU (part1)
1:49 min cleanup old file @ 2% CPU (part2)
Shutdown:
Checkpointer: 1%CPU 24:51

**Comment by pashkinelfe** [comment]
On the same s3 db as previous measurements:

1st Startup time:
53s counting partitions @100% CPU (part1)
1m:49s cleanup old file @ 2% CPU (part2)

1st Shutdown time 35 min:
Checkpointer: **1%**CPU
![shutdown-s3-10000-checkpointer-new](https://github.com/orioledb/orioledb/assets/63344111/b8136bb6-b93b-44f7-b691-c1f026f3bdaf)
S3 workers (2%-7%) each
![shutdown-s3-10000-s3worker-new](https://github.com/orioledb/orioledb/assets/63344111/100ce41e-34b9-4bcc-b909-13aad0ef374a)
Messages like 
```
s3_header_mark_part_loading(5 29908 5 0 0) - 12638
2024-01-04 10:18:11.450 UTC [35049] DEBUG:  S3 map put orioledb_data/13/5/29902.map orioledb_data/5/29902-13.map
2024-01-04 10:18:11.479 UTC [35040] DEBUG:  S3 part get orioledb_data/5/5/29908.0.0 orioledb_data/5/29908-5
2024-01-04 10:18:11.577 UTC [35056] DEBUG:  S3 map put orioledb_data/13/5/29908.map orioledb_data/5/29908-13.map
2024-01-04 10:18:11.578 UTC [35037] DEBUG:  s3_header_mark_part_loading(5 29920 5 0 0) - 12639
``` 
In the log. (Maybe effect of https://github.com/orioledb/orioledb/commit/be17f3abdff23f2e27f0d9bea6e2b175d651df72)

2nd Startup time:
6s counting partitions @100% CPU (part1)
52s cleanup old file @ 2% CPU (part2)

2nd Shutdown time
<waitiing for results>

**Comment by pashkinelfe** [comment]
Wallclock (gdb) flamegraph same as https://github.com/orioledb/orioledb/issues/328#issuecomment-1870347015
taken at be17f3abdff23f2e (last commit without segfault #333)
start1 (counting partitions @100%)
![start1](https://github.com/orioledb/orioledb/assets/63344111/81637b00-c979-4d71-bbc9-a4b3b4cc276d)
start2 (cleanup @2%)
![start2](https://github.com/orioledb/orioledb/assets/63344111/39f3dbae-a83e-4d04-8280-a9f7864642df)

stop (checkpointer @1%)
![stop-chkp](https://github.com/orioledb/orioledb/assets/63344111/0dbbe2d0-c6f6-4faa-8f1b-4a9c583ac047)
stop (s3 worker @30%)
![stop-s3w2](https://github.com/orioledb/orioledb/assets/63344111/d8fd17f9-d25e-45d4-9af7-4c20c9b0b90c)

bgworker when at @100%
![bgw](https://github.com/orioledb/orioledb/assets/63344111/4f82cf81-87e3-480f-8e49-dfdca1cf8df7)

non-recompressed svg:
[wallclock-flamegraph-s3-10000.zip](https://github.com/orioledb/orioledb/files/13900952/wallclock-flamegraph-s3-10000.zip)

**Comment by akorotkov** [comment]
@pashkinelfe, please, check after 234115c there is no function verify_dir_is_empty_or_create() (on the flamegraph), instead there is verify_dir_exists_or_create().

**Comment by pashkinelfe** [comment]
https://github.com/orioledb/orioledb/commit/234115c782dd86b910a87cfd782bbec18d2f9361 experience segfault https://github.com/orioledb/orioledb/issues/333 so the cluster could not be filled with data in a reproducible way. I'd suggest fixing #333 then recheck

**Comment by pashkinelfe** [comment]
at [02bc078](https://github.com/orioledb/orioledb/commit/02bc078a30f71940500be7eb63ff86badb17a3fa)
startup: 60s (counting partitions)
shutdown 10s
